src/inference/graph_inference.cpp

```diff
@@ -786,6 +879,137 @@ class GraphNodeAssignment : public Nice2Assignment {
+      double best_score = 0;
+      std::vector<int> best_assignments;
+      for (auto var = inf_vars.begin(); var != inf_vars.end(); var++) {
+        best_score += GetNodeScore(fweights, *var);
+        best_assignments.push_back(assignments_[*var].label);
+      }
+      std::vector<Factor> factors_candidates;
+      for (uint j = 0; j < factors.size(); j++) {
+        int factor_matches_giv_vars = true;
+        for (auto label = giv_labels.begin(); label != giv_labels.end(); label++) {
+          if (factors[j].count(*label) == 0) {
+            factor_matches_giv_vars = false;
+            break;
+          }
+        }
+        if (factor_matches_giv_vars) {
+          factors_candidates.push_back(factors[j]);
+        }
+      }
+      for (uint j = 0; j < factors_candidates.size(); j++) {
+        Factor current_candidate = factors_candidates[j];
+        std::set<int> distinct_labels;
+        std::vector<int> candidate_inf_labels;
+        std::unordered_set<std::vector<int>> permutations;
+        for (auto label = current_candidate.begin(); label != current_candidate.end(); label++) {
+          if (giv_labels.count(*label) == 0) {
+            candidate_inf_labels.push_back(*label);
+            distinct_labels.insert(*label);
+          }
+        }
+        uint64 num_permutations = CalculateFactorial(distinct_labels.size());
+        if (distinct_labels.size() > 64 || num_permutations > FLAGS_permutations_beam_size) {
+          RandomPermute(candidate_inf_labels, permutations);
+        } else {
+          HeapPermute(candidate_inf_labels, permutations, candidate_inf_labels.size());
+        }
+        for (auto it = permutations.begin(); it != permutations.end(); it++) {
+          for (uint z = 0; z < inf_vars.size(); z++) {
+            assignments_[inf_vars[z]].label = (*it)[z];
+          }
+
+          bool is_assignment_valid = true;
+          for (uint z = 0; z < inf_vars.size(); z++) {
+            if (HasDuplicationConflictsAtNode(inf_vars[z]) ||
+                !fweights.label_checker_.IsLabelValid(assignments_[inf_vars[z]].label)) {
+              is_assignment_valid = false;
+              break;
+            }
+          }
+          if (is_assignment_valid == false) {
+            continue;
+          }
+          double score = 0;
+          std::vector<int> assignment;
+          for (uint z = 0; z < inf_vars.size(); z++) {
+            score += GetNodeScore(fweights, inf_vars[z]);
+            assignment.push_back(assignments_[inf_vars[z]].label);
+          }
+          if (score > best_score) {
+            best_assignments = assignment;
+            best_score = score;
+          }
+        }
+      }
+      for (uint j = 0; j < inf_vars.size(); j++) {
+        assignments_[inf_vars[j]].label = best_assignments[j];
+      }
+    }
+  }
+
+  void HeapPermute(std::vector<int> v, std::unordered_set<std::vector<int>>& permutations, int n) {
```

Comment:
No it is just the name of the algorithm that I found on the internet, I will call it  ComputeAllPermutations